src/api/posts/routes.py

```diff
@@ -19,16 +24,16 @@
 
 @topics_ns.route("/<int:topic_id>/posts")
 class PostsList(Resource):  # type: ignore
-    """Get a list of posts of a certain topic."""
+    """Work with posts."""
 
     @staticmethod
     @authorized_access()
     def get(topic_id: int) -> list[dict[str, Any]] | None:
         """Get a list of posts of a certain topic."""
-        parser.add_argument("order_by", type=parse_order_by, help="get a list of posts of a topic")
-        sorting_parameters = parser.parse_args()
```

Comment:
Why did we change these lines?

Reply:
I moved topic validation higher because it kinda made sense. It seems rational to check for actual post before executing any  additional code

Reply:
Yes, I understand your motivation, It really may seem like we need to do full validation from left to right.
For example for this route:
```http
GET /topics/42/posts/18
```
it may seem that first we need to validate `/topics/42/` part at first, and only if it's correct, check the other part `posts/18`. And it is the right intention, but in reality we need to do **formal** validation of the **whole** request and only after that execute any other logic because of the application layers and responsibility separation. 

Our application has leveled architecture: request parsing/validation > controller logic > db access > response serialization. This means that we need to do everything possible with request validation first, without any db queries. That's why I used word **formal** in case of request validation. Because from some point of view, we can treat checking topic existence as a validation of `/topics/42` route part and it actually is, but it is **logic** validation, and this type of validation should be done on the next layer. First layer should be responsible only for **formal** validation, and do as much as possible for that.

So in layered architecture we should parse and check request **formaly** first. We should check that it contains allowed words, required parameters, correctness of parameter data types, and so on. After we checked the whole request for its correct format, we can go to the next app layer and check request **logically**, via access to db or other internal/external services or application logic itself.

And also worth to mention that actually a Flask (Flask-RESTx) does a great job to confuse you in this case. In many frameworks request parsing and validation by design is not a part of the route function body. For example, in FastAPI request validation is done in Pydantic schemas and decorators. In Django it's done in DRFModels (I don't remember the actual name, but the idea is correct).

Actually we could also write a decorator which accepts reqparse and provides parsed data to the route.

Reply:
Yes, I understand your motivation, It really may seem like we need to do validation from left to right.
For example for this route:
```http
GET /topics/42/posts/18
```
it may seem that first we need to validate `/topics/42/` part, and only if it's correct, check the other part `posts/18`. And it is the right intention, but in reality we need to do **formal** validation of the whole request and only after that execute any other logic because of the application layered architecture and responsibility separation. 

Our application has layered architecture: request parsing/validation > controller logic > db access > response serialization. This means that we need to do everything possible with request validation first, without any db queries. That's why I used word **formal** in case of request validation. Because from some point of view, we can treat checking topic existence as a validation of `/topics/42` route part and it actually is, but it is a **logic** validation, and this should be done on the next layer. First layer should be responsible only for **formal** request validation, and do as much as possible for that.

So for layered architecture we should parse and check request **formaly** first. We should check that it contains allowed words, required parameters, correctness of parameter data types, and so on. After we checked the whole request for its correct format, we can go to the next app layer and check request **logically**, via access to db or other internal/external services or application logic itself.

And also worth to mention that actually a Flask (Flask-RESTx) does a great job to confuse you in this case. In many frameworks request parsing and validation by design is not a part of the route function body. For example, in FastAPI request validation is done in Pydantic schemas and decorators. In Django it's done in DRFModels (I don't remember the actual name, but the idea is correct).

Actually we could also write a decorator which accepts reqparse and provides parsed data to the route.